module/zfs/spa.c

```diff
@@ -5890,9 +5899,11 @@ spa_export_common(char *pool, int new_state, nvlist_t **oldconfig,
 		if (!hardforce)
 			spa_write_cachefile(spa, B_TRUE, B_TRUE);
 		spa_remove(spa);
+	} else {
+		spa->spa_exporting = B_FALSE;
```

Comment:
I may be mistaken about this, but I think we need to always set `spa->spa_exporting = B_FALSE;` when returning `0`.  Otherwise, if the pool is exported  but not removed from the namespace, for example by `spa_reset()`.  Then if it's re-imported I don't believe we'll be able to export it.

Reply:
My mistake, this looks good since the else clause here is specifically for `spa_reset()` otherwise the `spa_t` is freed.

Reply:
Per our discussion I added a generic comment